src/js/protocols/webusbdfu.js

```diff
@@ -0,0 +1,1119 @@
+        .reset()
+        .then(() => {
+            console.log('Reset Device');
+            callback?.();
+        })
+        .catch(error => {
+            console.log(`Could not reset device: ${error}`);
+            callback?.();
+        });
+    }
+    getString(index, callback) {
+        const setup = {
+            requestType: 'standard',
+            recipient: 'device',
+            request: 6,
+            value: 0x300 | index,
+            index: 0, // specifies language
+        };
+
+        this.usbDevice
+        .controlTransferIn(setup, 255)
+        .then(result => {
+            if (result.status === 'ok') {
+                const _length = result.data.getUint8(0);
+                let _descriptor = '';
+                for (let i = 2; i < _length; i += 2) {
+                    const charCode = result.data.getUint16(i, true);
+                    _descriptor += String.fromCharCode(charCode);
+                }
+                callback(_descriptor, 0);
+            } else {
+                throw new Error(`USB getString failed! ${result.status}`);
+            }
+        })
+        .catch(error => {
+            console.log(`USB getString failed! ${error}`);
+            callback('', 1);
+        });
+    }
+    getInterfaceDescriptors(interfaceNum, callback) {
+        const self = this;
+
+        let interfaceID = 0;
+        const descriptorStringArray = [];
+        const interfaceCount = self.usbDevice.configuration.interfaces.length;
+        let descriptorCount = 0;
+
+        if (interfaceCount === 0) {
+            callback(0, 1); // no interfaces
+        } else if (interfaceCount === 1) {
+            descriptorCount = self.usbDevice.configuration.interfaces[0].alternates.length;
+        } else if (interfaceCount > 1) {
+            descriptorCount = interfaceCount;
+        }
+
+        const getDescriptorString = function () {
+            if (interfaceID < descriptorCount) {
+                self.getInterfaceDescriptor(interfaceID, function (descriptor, resultCode) {
+                    if (resultCode) {
+                        callback([], resultCode);
+                        return;
+                    }
+                    interfaceID++;
+                    self.getString(descriptor.iInterface, function (descriptorString, resultCode) {
+                        if (resultCode) {
+                            callback([], resultCode);
+                            return;
+                        }
+                        if (descriptor.bInterfaceNumber === interfaceNum) {
+                            descriptorStringArray.push(descriptorString);
+                        }
+                        getDescriptorString();
+                    });
+                });
+            } else {
+                //console.log(descriptorStringArray);
+                callback(descriptorStringArray, 0);
+                return;
+            }
+        };
+        getDescriptorString();
```

Comment:
```suggestion
        let interfaceID = 0;
        const descriptorStringArray = [];
        const interfaceCount = this.usbDevice.configuration.interfaces.length;
        let descriptorCount = 0;

        if (interfaceCount === 0) {
            callback(0, 1); // no interfaces
        } else if (interfaceCount === 1) {
            descriptorCount = this.usbDevice.configuration.interfaces[0].alternates.length;
        } else if (interfaceCount > 1) {
            descriptorCount = interfaceCount;
        }

        const getDescriptorString = () => {
            if (interfaceID < descriptorCount) {
                this.getInterfaceDescriptor(interfaceID, function (descriptor, resultCode) {
                    if (resultCode) {
                        callback([], resultCode);
                        return;
                    }
                    interfaceID++;
                    this.getString(descriptor.iInterface, (descriptorString, resultCode) => {
                        if (resultCode) {
                            callback([], resultCode);
                            return;
                        }
                        if (descriptor.bInterfaceNumber === interfaceNum) {
                            descriptorStringArray.push(descriptorString);
                        }
                        getDescriptorString();
                    });
                });
            } else {
                callback(descriptorStringArray, 0);
                return;
            }
        };
        getDescriptorString();
```

Reply:
Done. Line 225 should be:

`this.getInterfaceDescriptor(interfaceID, (descriptor, resultCode) => {`

Reply:
Using this here seems to break flashing. So reverted for now.